### ISSUE TITLE
How should training be changed when using image augmentation?

### ISSUE THREAD

**Issue by patrickcgray** [matterport/Mask_RCNN#1015]
I've got a nice augmentation setup using the great imgaug library that was chosen for this project but I'm not certain how I should change my training parameters given this augmentation.

I'm training basically using the same example from the imgaug documentation with some minor changes:
```
seq = iaa.Sometimes(0.833, iaa.Sequential([
    iaa.Fliplr(0.5), # horizontal flips
    iaa.Crop(percent=(0, 0.1)), # random crops
    # Small gaussian blur with random sigma between 0 and 0.5.
    # But we only blur about 50% of all images.
    iaa.Sometimes(0.5,
        iaa.GaussianBlur(sigma=(0, 0.5))
    ),
    # Strengthen or weaken the contrast in each image.
    iaa.ContrastNormalization((0.75, 1.5)),
    # Add gaussian noise.
    # For 50% of all images, we sample the noise once per pixel.
    # For the other 50% of all images, we sample the noise per pixel AND
    # channel. This can change the color (not only brightness) of the
    # pixels.
    iaa.AdditiveGaussianNoise(loc=0, scale=(0.0, 0.05*255), per_channel=0.5),
    # Make some images brighter and some darker.
    # In 20% of all cases, we sample the multiplier once per channel,
    # which can end up changing the color of the images.
    iaa.Multiply((0.8, 1.2), per_channel=0.2),
    # Apply affine transformations to each image.
    # Scale/zoom them, translate/move them, rotate them and shear them.
    iaa.Affine(
        scale={"x": (0.8, 1.2), "y": (0.8, 1.2)},
        translate_percent={"x": (-0.2, 0.2), "y": (-0.2, 0.2)},
        rotate=(-25, 25),
        shear=(-8, 8)
    )
], random_order=True)) # apply augmenters in random order
```

So given that this code basically runs these augmentations randomly for 5 out of 6 images should I increase my steps per epoch x6? Should I increase my num of epochs x6? How does the augmentation actually get applied during training? 

I see it is applied in load_image_gt() which is called in the data_generator() but I'm not totally clear what should change in my training regimen based on these augmentations.

**Comment by nayzen** [comment]
Hi Patrick,

I'm trying to artificially augment my dataset to obtain better performances. I have VIA annotations (segmentations) and I want to use the Imgaug library but I don't know where to use it in the Mask R-CNN code. 
Moreover, my code works with annotations associated with each image, so I don't know if the segemntations can be augmented too ? :/

Regards,
Antoine

**Comment by patrickcgray** [comment]
Hi @nayzen you can see how you use imgaug in the coco example here: https://github.com/matterport/Mask_RCNN/blob/8bed8428a9698825de06ff7277fd5cff91aeb77f/samples/coco/coco.py#L494

I'm not exactly sure what you mean by the segmentations being augmented too and if your code is different than normal, but this implementation is written so that all labels (aka segmentations aka polygons) will be augmented along with the image. Otherwise it wouldn't do anyone much good. So I would assume that you can just add a bunch of augmentations and it will work fine!

**Comment by nayzen** [comment]
Hi @patrickcgray , and thank you for your quick answer !
I didn't check this part of code and I'm glad you shared it with me. By segmentations I mean polygons (made with the VIA tool), I have a .JSON file with polygons coordinates for each image, and I was wondering how to augment these annotations AND the images at the same time. But I think you're right :)

![example_segmaps](https://user-images.githubusercontent.com/37196095/51167323-98e08700-18a6-11e9-9c06-2c078e86aa3d.jpg)

This example is exactly what I want.

**Comment by patrickcgray** [comment]
@nayzen perfect! And very cool image demonstrating it!

**Comment by ash1995** [comment]
I'm confused as to how the flipping image augmentation works. So say I had a building in the left side of the image that I wanted to detect, flipping it horizontally would result it in being located on the right side. But the training annotation will state that the building is on the left side. Won't this confuse the network. My question is how is this accounted for?

**Comment by patrickcgray** [comment]
@ash1995 look at the post @nayzen just made. You can see the annotations are augmented in the same way as the image so this is accounted for internally.

**Comment by nayzen** [comment]
Hello @ash1995 , I understand your confusion. I created an issue here :
https://github.com/aleju/imgaug/issues/234
where I'm asking how augmentation works for the annotations if we apply geometric transformations.
In my case, I have to convert my polygons to keypoints, and then use a function to augment the images AND the annotations.
Because, without geometric transformation you can augment your dataset without augmenting your annotations.
Hope it'll help

**Comment by ash1995** [comment]
So if I were to do apply the **flip** augmentation like flipping the image left to right or upside down, I don't need to change the annotation. But if I applied a geometric augmentation like rotation, I will need to change the annotation as discussed in the issue you referenced: [aleju/imgaug#234](https://github.com/aleju/imgaug/issues/234) 

Am I understanding this correctly?

**Comment by nayzen** [comment]
@ash1995 Flip augmentation (right, up and down) are geometric transformation too, so you cannot do that.
In my post above, it was an example from the imgaug doc. It shows that annotations are also augmented, and that's what I'm looking for. 
For the moment I augmented my database without geometric transformations as below:

![imgaug_2](https://user-images.githubusercontent.com/37196095/51516929-5a544a80-1e19-11e9-9f5d-934f7984e45f.png)

**Comment by ash1995** [comment]
Ah I see. But matterport's implementation of Mask RCNN does use the flip augmentation, the Fliplr to be specific and it seemingly works which was what led me to believe that the flip augmentations work within the model.

**Comment by nayzen** [comment]
Yes I saw but I specified my own transformations just in case.

**Comment by ash1995** [comment]
Ok, so as I understand it, augmentations that do not change the orientation of the images are definitely safe to use. Matterport uses the Fliplr augmentation so it's most probably safe to use as well. Same goes for the Flipud. I'm working on satellite images and so I do need the flip augmentations.

**Comment by nayzen** [comment]
You can try doing fliplr /ud and tell me if it works for you ! :-)

**Comment by ash1995** [comment]
Will do. Thanks a lot @nayzen

**Comment by patrickcgray** [comment]
Hey @ash1995 and @nayzen, I suppose I just assumed this would work. I've been using all the augmentations described in my [first post](https://github.com/matterport/Mask_RCNN/issues/1015#issue-368371797) and it seems to work well  based on training/validation loss but I haven't actually verified it. How are you checking to see if the augmentations are being applying to the annotations?

**Comment by ash1995** [comment]
@patrickcgray  @nayzen 

Ok, I think there isn't a problem with augmentation. The reason this is so is because the bounding box and segmentation masks used within the project is not from the annotations but is calculated internally. The workflow for loading an image is as follows:

1. Get image and its corresponding segmentation mask coordinates & class_ids from annotations
2. Resize image to the size defined in config
3. Modify masks **if** augmentation is specified
4. Extract bounding boxes from the masks

This is a screenshot of the load_image_gt in model.py that illustrates the workflow.
![git](https://user-images.githubusercontent.com/32295036/51595371-05dfc680-1f10-11e9-923a-a1d59647e608.png)


So we need not worry about changing the annotations because it's not using the annotations internally.

**Comment by nayzen** [comment]
Hi @patrickcgray , actually I didn't try to do geometric transformations for the moment.
But I saw a difference doing basic transformations (contrast/blur/channelShuffle) with my loss curve. 
It took more time to decrease than before, and I have better results now.

![before_after](https://user-images.githubusercontent.com/37196095/51661178-5b6bb000-1fb0-11e9-8194-56c351ecc43c.PNG)

My object was too bright and not detected. After image augmentation it is now :-)

@ash1995 I'm planning to try with geometric transformations later, because I also check this part of code and I think it could work too :-)

**Comment by ectg** [comment]
Hi, I was just trying to apply contrast, crop and blur augmenters to training, and I wasn't quite sure of the usage, especially with keywords such as _SomeOf_, _OneOf_ and _Sometimes_. 

If I say, `aug = iaa.OneOf([iaa.ContrastNormalization((0.5, 1.5)), iaa.GaussianBlur(sigma=(0.0, 5.0))])` is either blur or contrastNorm applied to every image in the training set in all epochs? If the original training set contains 100 images, does this augmentation increase the size to 200 images?

**Comment by ash1995** [comment]
1) `aug = iaa.OneOf([iaa.ContrastNormalization((0.5, 1.5)), iaa.GaussianBlur(sigma=(0.0, 5.0))])` as the name suggests will only apply one of the augmentations to a particular image. Refer the documentation [(link)](https://imgaug.readthedocs.io/en/latest/source/augmenters.html?highlight=OneOf#oneof) for an example.
2) Augmentation does not increase the size of the training dataset.

**Comment by ectg** [comment]
Hi @ash1995 , thank you. As to #2, if I use OneOf, is one of those augmentations applied to every image in the training set? And this would mean that the original unaugmented images are never seen by the network?

**Comment by ash1995** [comment]
Yes. So if your augmentation was:
 `aug = iaa.OneOf([iaa.ContrastNormalization((0.5, 1.5)), 
                               iaa.GaussianBlur(sigma=(0.0, 5.0))])` your images will either have a gaussian blur or will be subjected to contrast normalisation. No unaugmented image will be shown to the network.

**Comment by ectg** [comment]
Hi @ash1995, thanks! 

Is there any way to ensure that all unaugmented images are shown to the network? In addition to any augmentations I choose. 

The other option is to pre-augment the training set (save the augmentations as .png files) instead of doing augmentations on the fly, just want to be sure before I do this.

**Comment by ash1995** [comment]
If you're going with the imgaug library then no, I dont't think you can show the entire unaugmented dataset **and** the augmented dataset together. I believe imgaug has a None augmentation wherein it won't apply any augmentation to an image but that's just for that particular subset of images to which you apply it to. If you truly want to double your dataset i.e. unaugmented + augmented datasets, then you'll most likely have to pre-augment the training set as you said.

**Comment by h-ann** [comment]
Hi @ectg ,
If you want to show both unaugmented and augmented images to network, you could use 
`iaa.SomeOf((0, 1), [iaa.ContrastNormalization((0.5, 1.5)), iaa.GaussianBlur(sigma=(0.0, 5.0))]`
and this will apply 0 to 1 given augmenters.
https://imgaug.readthedocs.io/en/latest/source/augmenters.html#someof - to read about this

**Comment by DerOzean** [comment]
Hi Everyone,

I try to add `ChangeColorspace` to my augmentation to get a better hue value:
[(http://learn.leighcotnoir.com/artspeak/elements-color/hue-value-saturation/)]
The problem is that masks don't have the color space of an RGB and therefore I get an error. 
This is the code I want to add: 
`iaa.Sometimes(0.25,# change the colorspace from RGB to HSV, then add 50-100 to the first channel, then convert back to RGB. This increases the hue value of each image
                iaa.ChangeColorspace(from_colorspace="RGB", to_colorspace="HSV"),
                iaa.WithChannels(0, iaa.Add((50, 100))),
                iaa.ChangeColorspace(from_colorspace="HSV", to_colorspace="RGB")
),`

My Question: Is there a way to add some changes only to the images and not to the masks?
It would be great just to change some images from RGB to HSV.

**Comment by blimeyitscode** [comment]
@patrickcgray @nayzen Thanks a lot for your combined contributions, I was wondering if there was a way to train the Matterport algorithm on image masks themselves without need for a Json? so the training set comprised of image masks instead of annotated Json information being fed in?

**Comment by nayzen** [comment]
Hi @gaurang84 , I always worked with Json coordinates (VIA tool), therefore I didn't try without them.

**Comment by rupa1118** [comment]
Hello everyone, I have a doubt please me in help understanding...
1. if we apply image augmentation on training dataset then the number of augmentation applied must be multiplied by the training set wt I mean is if say we apply fliplr and flipud then training set = actual training set *3 as we have original training set ??? Is this true ??

2. Should we need to annotate the augmented images ???  Or its not required as these are augmented images so the annotations of actual training set will be used here ????

3. Steps per epoch can be default 100 no matter what the training set size ??? Because many used the default steps per epoch. But in traditional approach steps per epoch = no of training samples / batch size. So what must be the value of steps per epoch ???

**Comment by kimile599** [comment]
> Hello everyone, I have a doubt please me in help understanding...
> 
> 1. if we apply image augmentation on training dataset then the number of augmentation applied must be multiplied by the training set wt I mean is if say we apply fliplr and flipud then training set = actual training set *3 as we have original training set ??? Is this true ??
> 2. Should we need to annotate the augmented images ???  Or its not required as these are augmented images so the annotations of actual training set will be used here ????
> 3. Steps per epoch can be default 100 no matter what the training set size ??? Because many used the default steps per epoch. But in traditional approach steps per epoch = no of training samples / batch size. So what must be the value of steps per epoch ???

For the third question, you can enlarge the steps per epoch to reduce the frequency of weights saving which will save you some space when you are working with colab.

**Comment by FloBr49** [comment]
Hi everyone,

i have to prepare a project for college and i'm tryin to do it with matterport Mask RCNN. My problem is now that i have images with lots of annotations (around 25 annotations with 8 different classes). To collect enough annotated data is too time consuming. So my question is: How do i use the image augmentation, especially geometric transformation, to enlarge my training dataset and automatically transform my annotations? 

You can find an example JSON file I want to use attached.

Thank you in advance.

[TEST_JSON_FB.zip](https://github.com/matterport/Mask_RCNN/files/6353272/TEST_JSON_FB.zip)

**Comment by patrickcgray** [comment]
@FloBr49 take a look at at the post @nayzen made here https://github.com/matterport/Mask_RCNN/issues/1015#issuecomment-454303949 that and the above comments show where to make the augmentations and what libraries to use. I've tested these and they all work well for augmenting a dataset.

**Comment by mansi-aggarwal-2504** [comment]
Hello! I am training on 150 (train) + 50 (val) images. Epochs = 70 and steps per epoch = 50. I am using Google Colab for training. My augmentations:

```
augmentation = iaa.SomeOf((0, None), [
iaa.Fliplr(1),
iaa.Flipud(1),
iaa.OneOf([iaa.Affine(rotate = 90 * i) for i in range(1, 3)]),
iaa.OneOf([iaa.Affine(rotate = -90 * i) for i in range(1, 3) ]),
iaa.Affine(scale={"x": (0.5, 1.5), "y": (0.5, 1.5)}),
iaa.sizelib.CropAndPad(percent=([0.05, 0.1])),
])
```
The training would always take 7-8 hours but now, it seems stuck. 
<img width="597" alt="Screenshot 2021-07-27 at 2 51 33 PM" src="https://user-images.githubusercontent.com/60866597/127129945-acd124f6-3304-49be-9eb0-219094391062.png">

How much time does training usually take with augmentations on a data of this size? Any inputs?

**Comment by raficabral** [comment]
> @patrickcgray @nayzen Thanks a lot for your combined contributions, I was wondering if there was a way to train the Matterport algorithm on image masks themselves without need for a Json? so the training set comprised of image masks instead of annotated Json information being fed in?

Hi @blimeyitscode, I am wondering the same question. Did you handle how to train with matterport framework without need for a Json?
